docs/developer/decision-records/2024-10-09-vault-authentication-refactor/README.md

```diff
@@ -0,0 +1,223 @@
+
+    public VaultAuthenticationRegistryImpl() {
+    }
+
+    @Override
+    public void register(String context, HashicorpVaultAuth service) {
+        services.put(method, service);
+    }
+
+    @Override
+    public @NotNull HashicorpVaultAuth resolve(String method) {
+        return services.get(method);
+    }
+
+    @Override
+    public boolean hasService(String method) {
+        return services.containsKey(method);
+    }
+}
+```
+
+### Registration
+
+Each Service Implementation will register itself in the `VaultAuthenticationRegistry`.
+This is done through an `@Provider` method inside the HashiCorp Vault extension, that is then injected with `@Inject` into the Service Implementation.
+
+```java
+@Provider
+public VaultAuthenticationRegistry vaultAuthenticationRegistry() {
+    return new VaultAuthenticationRegistryImpl();
+}
+```
+
+The `VaultAuthenticationRegistry` is injected into the Vault Authentication Service Implementations through `@Inject`.
+
+```java
+@Inject
+private VaultAuthenticationRegistry vaultAuthenticationRegistry;
+```
+
+It will then be registered during initialization through:
+
+```java
+@Override
+public void initialize(ServiceExtensionContext context) {
+    var token = context.getSetting(VAULT_TOKEN, null);
+
+    if (!vaultAuthenticationRegistry.hasService("token-based")) {
+        vaultAuthenticationRegistry.register("token-based", HashicorpVaultTokenAuth(token));
+    }
+
+}
+```
+
+### Configuration
+
+A new config value is introduced to the HashiCorp Vault Extension named `edc.vault.hashicorp.auth.method`. This value governs which Vault Authentication Service Extension is used from the Vault Auth Registry.
+
+Due to the possible differences in the configuration of each authentication method, each Vault Authentication Service Extension comes with its own configuration values.
+
+### HashiCorp Vault Client
+
+Since the client contains a lot of authentication logic, it will also go through a refactoring.
+After the refactoring it will only use the `VaultAuthenticationRegistry` based on the configured `edc.vault.hashicorp.auth.method` to fetch the `client_token` that is provided by the Vault Authentication Service Extensions.
+The `client_token` will then be used to generate the Headers for the HTTP requests.
+
+Old `getHeaders()`:
+
+```java
+@NotNull
+private Headers getHeaders() {
+    var headersBuilder = new Headers.Builder().add(VAULT_REQUEST_HEADER, Boolean.toString(true));
+    headersBuilder.add(VAULT_TOKEN_HEADER, settings.token());
+    return headersBuilder.build();
+}
+```
+
+Example for new `getHeaders()`:
+```java
+@NotNull
+private Headers getHeaders() {
+    var headersBuilder = new Headers.Builder().add(VAULT_REQUEST_HEADER, Boolean.toString(true));
+    headersBuilder.add(VAULT_TOKEN_HEADER, vaultAuthenticationRegistry.resolve(settings.getAuthMethod()).getVaultToken());
+    return headersBuilder.build();
+}
+```
+
+### HashiCorp Vault Token Authentication Extension
```

Comment:
`spi` is not mentioned here, but it will be necessary to provide the interfaces for the eventual extensions

Reply:
the addition of an `spi` module for the authentication already has a mention in its own section `Vault Auth SPI`